Adjust overlay header elements to match osu-web

thewildtree opened the pull request:
Sizing of the `OverlayHeader` and its elements was quite off from what can be seen on osu-web. This PR brings its values as close to the web counterpart as possible.

master:
![image](https://user-images.githubusercontent.com/27722894/75286713-3b449b00-5819-11ea-9553-b5d764ea6b93.png)


this PR:
![image](https://user-images.githubusercontent.com/27722894/75634146-488fca00-5c0b-11ea-9761-b1af1e827931.png)


web:
![image](https://user-images.githubusercontent.com/27722894/75634027-46793b80-5c0a-11ea-8e04-9b2478a4c2c9.png)



Please keep in mind that this PR might make things look a bit *worse* for now since it still needs a font sizing fix to look exactly like osu-web. Until then, to mantain proper sizing and prevent things from looking *too* off, I added line-height compensations (margins) to some of the text elements (similar to the ones from `BeatmapSetOverlay` resize PR) that will be safe to remove when the font sizing fix comes.

That being said, here's a comparison with a temporary font sizing fix applied (web on the left, lazer on the right):

![image](https://user-images.githubusercontent.com/27722894/75285014-1bf83e80-5816-11ea-963f-15a9d05d832d.png)

thewildtree committed:
Adjust OverlayRulesetTabItem font size to match web

thewildtree committed:
Adjust ScreenTitle to match osu-web

thewildtree committed:
Separate OverlayTabControl text colour from accent (bar) colour

thewildtree committed:
Adjust OverlayHeaderBreadcumbControl text colour

thewildtree committed:
Adjust sizing and spacing in OverlayHeader

TheWildTree commented on osu.Game/Overlays/OverlayTabControl.cs:
```diff
@@ -32,6 +33,20 @@ public override Color4 AccentColour
+                    i.TextColour = value;
+            }
+        }
+
```
I had to separate the text colour from the bar colour, but I'm not sure if this is the right way to do it.

bdach replied:
Ideally this shouldn't exist.

bdach replied:
By the way, since that's important - *why* is it necessary exactly to separate the colours? Because if I do things with `AccentColour` then it'll probably be necessary to consider how this should behave too.

TheWildTree replied:
See osu-web: in header tabs, separate colours are used for the text and for the bar below it. Previous lazer implementation didn't account for that and just used `AccentColour` for both.

Plus it's not consistent, header tabs use colours from the themed pallete while, for example, the profile section tabs just use plain white.

bdach replied:
Actually yeah, it has a chance of working. If the `IAccentColourProvider` is injectable it can also be provided in BDL which is like 100% better since then there's one central place for overall colouring. Or even we could entirely disregard it and just use `OverlayColourProvider` exclusively directly in the overlay tabs.

TheWildTree replied:
Second option sounds much easier (and actually doable in a few lines of code I think), that is if dropping `AccentColour` in its entirety despite deriving from `OsuTabControl` would be acceptable. If we went in that direction, we'd also have to make the bar below the tabs coloured in the same way, but that's a non-issue really.

bdach replied:
It's not a good idea to ignore the interface members. Way I see it `AccentColour` on the tab containers needs to be fixed anyway as it's broken. If we want to disregard it entirely here than some higher-order refactoring needs to be made (decouple `AccentColour` entirely by extracting another abstract tab control) but that might paradoxically create even more owrk.

thewildtree committed:
Make ScreenTitle elements protected and remove dependency

This removes the separator dot colouring for now because it conflicted with MultiHeaderTitle which doesn't exactly use the osu-web style.

thewildtree committed:
Bring MultiHeaderTitle back to old values

thewildtree committed:
Remove unneeded using directive

peppy commented:
`OverlayTabControl` is confusing to me. It's derived in two places:

- `OverlayHeaderTabControl`, which sets `TextColour` but *not* `AccentColour`
- `ProfileTabControl`, which sets `AccentColour` but *not* `TextColour`

Is this correct? Also, is there a reason the colour specifications can't be moved local to `OverlayTabControl` itself, if it is to be used in overlays (it could get the colour provider locally then).

thewildtree committed:
Add title separator colouring to OverlayHeader

TheWildTree commented:
It's confusing indeed, but kind of makes sense:

- `OverlayHeaderTabControl` is the one we're restyling here; it's the tab control directly under the screen title and its text colour is different from its accent (bar) colour, 
- `ProfileTabControl` is only used in the profile overlay as the section controller ([this one](https://user-images.githubusercontent.com/27722894/75374291-47d5fb80-58cc-11ea-88ec-31fa803f6c4f.png)), it doesn't need any restyling so no changes here. I'll rename it to better reflect what it does, though, as it confused me as well.

I haven't noticed that earlier though and accidentally applied changes to the `ProfileTabControl` as well, will fix.

thewildtree committed:
Simplify setter

thewildtree committed:
Separate new margin values from OverlayTabControl

thewildtree committed:
Bring ProfileTabControl text colour in line with web

thewildtree committed:
Rename variable to better reflect its purpose

thewildtree committed:
Move colour specifications to OverlayTabControl

thewildtree committed:
Remove unnecessary font weight change

thewildtree committed:
Adjust OverlayRulesetTabItem shadow

Only adjusts the colour - a blur radius adjustment is needed to match web 1:1, but it's a bit of a hassle to add while having very little visible effects, so let's leave it for now.

thewildtree committed:
Adjust header side margins

thewildtree committed:
Remove unused const

While this should be still used somewhere in the MultiHeaderTitle (as ScreenTitle uses 10 here), I think the 4px difference is acceptable considering that the multiplayer lobby screen is up for a redesign anyway.

TheWildTree commented:
Adjusted the content X-axis margin - osu-web uses 50 everywhere except for the user profile page (where it uses 70) while lazer used 70 in all cases. 

Moved the base constant to the `OverlayHeader` for potential reference from other components and added the ability to adjust that margin in case we ever need to set a different one (like in `UserProfileOverlay` now).

thewildtree committed:
Remove incorrect leftover value

thewildtree committed:
Increase ruleset selector spacing to match osu-web

thewildtree committed:
Fix OverlayTabControl bar offset

bdach commented on osu.Game/Graphics/UserInterface/ScreenTitle.cs:
```diff
@@ -72,27 +79,29 @@ protected ScreenTitle()
+                            Size = new Vector2(5),
                             Colour = Color4.Gray,
+                            Margin = new MarginPadding { Top = 3 } // compensation for osu-web using a font here making the circle appear a bit lower
                         },
```
I checked and it should be possible to fully mirror web implementation here:

```csharp
                        Separator = new OsuSpriteText
                        {
                            Anchor = Anchor.Centre,
                            Origin = Anchor.Centre,
                            Font = OsuFont.GetFont(size: 20),
                            Colour = Color4.Gray,
                            Margin = new MarginPadding { Vertical = text_padding },
                            Text = "\u2022"
                        },
```

TheWildTree replied:
For some reason I thought we can't use unicode chars like this here (don't know why, haven't even tried tbh), thanks

bdach commented on osu.Game/Overlays/OverlayHeader.cs:
```diff
@@ -15,9 +15,18 @@ public abstract class OverlayHeader : Container
 
+        public const float CONTENT_X_MARGIN = 50;
+
+        public virtual float HorizontalMargin
```
Should be `protected`, get-only and used directly in the ctor. Basically make `CONTENT_X_MARGIN` into a protected get-only property.

TheWildTree replied:
Sorry, I don't think I understood this properly, could you elaborate? 

I don't agree that `CONTENT_X_MARGIN` should be made into anything `protected` since it might be used by other components outside the `OverlayHeader` or any classes deriving from it (see current `UserProfileOverlay.CONTENT_X_MARGIN and how it's used by other parts of the overlay). Unless this can be achieved in a different way?

We also need the header to let us customize this value in case an overlay needs a different one - like the `UserProfileOverlay` does in this PR - that's why I made `HorizontalMargin` like this. I do indeed feel like there's a better way to do that, but I don't have any concrete ideas right now.

bdach replied:
OK, those consts can stay for now I suppose, although they feel super wrong. Here I mainly have an issue with exposition as public. IMO would be better to invert flow:

```csharp
protected virtual float HorizontalPadding => CONTENT_X_MARGIN;
```

and then instead at `content` creation pass `new MarginPadding { Horizontal = HorizontalPadding }`.

TheWildTree replied:
Changed, please check if it's better now.

bdach commented on osu.Game/Graphics/UserInterface/ScreenTitle.cs:
```diff
@@ -36,18 +37,24 @@ protected IconUsage Icon
+        {
+            get => Separator.Colour;
+            set => Separator.Colour = value;
         }
```
Do these need to be a thing? Can't we just inject `OverlayColourProvider` locally?

TheWildTree replied:
This one has to be a thing as long as `MultiHeaderTitle` exists in the current form since the multiplayer room screen doesn't have an OCP higher up in the hierarchy.

bdach replied:
Alright. Could be worse all things considered.

bdach commented on osu.Game/Overlays/OverlayRulesetTabItem.cs:
```diff
@@ -53,6 +53,8 @@ public OverlayRulesetTabItem(RulesetInfo value)
                         Anchor = Anchor.Centre,
                         Text = value.Name,
+                        Font = OsuFont.GetFont(size: 18),
+                        ShadowColour = new Color4(0, 0, 0, 0.75f)
```
Slightly more readable:
```suggestion
                        ShadowColour = Color4.Black.Opacity(0.75f)
```

bdach commented on osu.Game/Overlays/UserProfileOverlay.cs:
```diff
@@ -149,30 +151,32 @@ private void userLoadComplete(User user)
-                AccentColour = AccentColour
+                // This needs to be set here manually because TabControl attempts to propagate these values to its children before these tabs are added
+                AccentColour = AccentColour,
+                TextColour = TextColour
```
Can you explain further? I'm kind of confused at why this is necessary.

TheWildTree replied:
At the moment these tabs are created, its parent `TabControl` has already attempted to set its children `AccentColour` / `TextColour` like here:
https://github.com/ppy/osu/blob/6b284fb3e74fbfce50c96e1096563360b95ba9cd/osu.Game/Graphics/UserInterface/OsuTabControl.cs#L29-L37

 This happens because they're being set when the `TabControl` is created, while the tabs in `ProfileSectionTabControl` are added a bit later (when their data is loaded up from the API):

https://github.com/ppy/osu/blob/6b284fb3e74fbfce50c96e1096563360b95ba9cd/osu.Game/Overlays/UserProfileOverlay.cs#L139-L149

`AccentColour` was already being set this way here, so I just added the `TextColour` which acts pretty much in the same manner.

bdach replied:
That is immensely broken & yet another argument to just move `OverlayColourProvider` down to the tabs themselves.

TheWildTree replied:
I moved the text colour change down to the tab items but left `AccentColour` as is, since that's being set by the `OsuTabControl` higher up in the hierarchy and I don't think I want to change that.

bdach replied:
A half-change feels worse than the existing state of affairs to be honest. Let's bring this back as it was.

bdach commented on osu.Game/Overlays/OverlayTabControl.cs:
```diff
@@ -55,19 +79,23 @@ protected class OverlayTabItem : TabItem<T>, IHasAccentColour
+                    textColour = value;
                     updateState();
                 }
             }
```
Again, can we not apply these locally with `OverlayColourProvider`?

bdach commented on osu.Game/Overlays/OverlayTabControl.cs:
```diff
@@ -40,12 +56,20 @@ protected OverlayTabControl()
             AddInternal(bar = new Box
             {
                 RelativeSizeAxes = Axes.X,
+                Margin = new MarginPadding { Top = -1 },
```
Not sure what this is fixing but this is a hack. Probably not worth the one pixel.

TheWildTree replied:
Without this, the coloured bar is displayed below the background box:
![image](https://user-images.githubusercontent.com/27722894/75719963-ff587c80-5cd5-11ea-9587-95e4c5955d51.png)

while on web it's inside it:
![image](https://user-images.githubusercontent.com/27722894/75720073-21ea9580-5cd6-11ea-8c90-350343a21ea4.png)

bdach replied:
This is probably the proper way to get rid of this:
```csharp
            AddInternal(bar = new Box
            {
                RelativeSizeAxes = Axes.X,
                Height = 2,
                Anchor = Anchor.BottomLeft,
                Origin = Anchor.BottomLeft
            });
```

FWIW, same thing probably goes for the expanding bar below:
```csharp
                    Bar = new ExpandingBar
                    {
                        Anchor = Anchor.BottomCentre,
                        Origin = Anchor.Centre,
                        ExpandedSize = 7.5f,
                        CollapsedSize = 0,
                    },
```

TheWildTree replied:
The bar fix you suggested does work, but the `ExtendedBar` is still 1px too low using these values:

![image](https://user-images.githubusercontent.com/27722894/75802882-c1169800-5d7d-11ea-8479-08784acf2be6.png)

bdach replied:
I suppose we can bring back the -1 but I'd make the 2 height into a `public const EXPANDED_TAB_HEIGHT` and reference that instead by doing `-EXPANDED_TAB_HEIGHT / 2` so it's obvious what it does.

TheWildTree replied:
Except the line we're seeing is using `Height = 1` as seen here:

https://github.com/ppy/osu/blob/6b284fb3e74fbfce50c96e1096563360b95ba9cd/osu.Game/Overlays/TabControlOverlayHeader.cs#L76-L85

I'll change this to be more readable (only the `ProfileSectionTabControl` uses `Height = 2` because having it in the base class is a bit misleading.

bdach replied:
Yeah I got confused because I looked at `OverlayTabControl` instead of `OverlayHeaderTabControl` by mistake. Should be half regardless.

TheWildTree replied:
Half of 1 doesn't really help us here though.

bdach replied:
Does it not? Since sizes are floats I thought it might.

TheWildTree replied:
We need a `-1` margin, `-0.5` looks just ever-so-slightly off at higher UI scales:

![image](https://user-images.githubusercontent.com/27722894/75807297-684afd80-5d85-11ea-994f-a3d5a92b0de7.png) ![image](https://user-images.githubusercontent.com/27722894/75807329-7731b000-5d85-11ea-88ea-ce292b4c879f.png)
(-0.5 on the left)

bdach replied:
OK, sure.

bdach commented on osu.Game/Screens/Multi/Header.cs:
```diff
@@ -71,11 +72,20 @@ private void load(OsuColour colours)
 
         private class MultiHeaderTitle : ScreenTitle
         {
+            private const int text_offset = 2;
```
Single-use constant (can be inlined), and again, not sure why this is required.

TheWildTree replied:
This was copied over from the previous version of `ScreenTitle`, it was used there to align the text to match the desired design I suppose? Can remove if not necessary anymore.

bdach replied:
If nobody knows why it's there let's just remove it.

thewildtree committed:
Remove unneeded margin

thewildtree committed:
Make value more readable

thewildtree committed:
Remove MultiHeaderTitle customizations

This does make its design a little different than before, I think to an acceptable extent. The whole multi room is a design mess right now so I feel like it's gonna be reworked anyway sooner or later.

thewildtree committed:
Refactor text colour changing

Moved the colour assignments directly to the `OverlayTabItem`s. Renamed `TextColour` to `IdleColour` there because it was confusing in its previous state (if being set outside the base `OverlayTabItem`, it could've been easily confused with `Text.Colour` which would not work properly in this case).

thewildtree committed:
Properly adjust bar position

thewildtree committed:
Remove unnecessary using

thewildtree committed:
Remove unnecessary using directive

thewildtree committed:
Improve horizontal margin setting

thewildtree committed:
Make bar height assignments less confusing

thewildtree committed:
Bring back text colour assignment to CreateTabItem

thewildtree committed:
Remove unused variable

thewildtree committed:
Rearrange extended bar size assignments

thewildtree committed:
Add temporary margin values

These should be replaced by constants (still have to figure out the proper way to do that).

bdach commented:
It's become apparent that `OsuTabControl`'s `AccentColour` brokenness and the overall byzantine hierarchy of the overlay tab controls needs urgent addressing, so I'm marking that as blocked for now and will be attempting to slowly fix it over the coming days.

TheWildTree commented:
Closing for now, I'll split changes not related to the `TabControl` mess into different PRs to further work on them (design changes, mostly).
